sematic/examples/testing_pipeline/pipeline.py

```diff
@@ -0,0 +1,197 @@
+    time.sleep(5)
+    m = []
+    while True:
+        m.append(" " * 2**10)
+    return val
+
+
+@sematic.func(inline=False)
+def do_raise(val: float) -> float:
+    """
+    Raises a ValueError, without retries.
+    """
+    logger.info("Executing: do_raise(val=%s)", val)
+    time.sleep(5)
+    raise ValueError("test error")
+
+
+@sematic.func(
+    inline=False, retry=sematic.RetrySettings(exceptions=(ValueError,), retries=10)
+)
+def do_retry(val: float, failure_probability: float = 0.5) -> float:
+    """
+    Raises a ValueError with the given probability, with a total of 10 retries.
+    """
+    logger.info(
+        "Executing: do_retry(val=%s, failure_probability=%s)", val, failure_probability
+    )
+    time.sleep(5)
+    if random.random() < failure_probability:
+        raise ValueError("test retriable exception")
+    return val
+
+
+@sematic.func(inline=True)
+def testing_pipeline(
+    inline: bool = False,
+    nested: bool = False,
+    fan_out: int = 0,
+    should_raise: bool = False,
+    raise_retry_probability: Optional[float] = None,
+    oom: bool = False,
+    exit_code: Optional[int] = None,
+) -> float:
+    """
+    The root function of the testing pipeline.
+
+    Its parameters control the actual shape of the pipeline, according to testing needs.
+
+    Parameters
+    ----------
+    inline: bool
+        Whether to include inline functions in the pipeline. Defaults to False.
+    nested: bool
+        Whether to include nested functions in the pipeline. Defaults to False.
+    fan_out: int
+        How many dynamically-generated functions to add in parallel. Defaults to 0.
+    should_raise: bool
+        Whether to include a function that raises a ValueError, without retries.
+        Defaults to False.
+    raise_retry_probability: Optional[float]
+        If not None, includes a function which raises a ValueError with the given
+        probability, with a total of 10 retries. Defaults to None.
+    oom: bool
+        Whether to include a function that causes an Out of Memory error.
+        Defaults to False.
+    exit_code: Optional[int]
+        If not None, includes a function which will exit with the specified code.
+        Defaults to None.
+    """
+    # have an initial function whose output is used as inputs by all other functions
+    # this staggers the rest of the functions and allows the user a change to monitor and
+    # visualize the unfolding execution
+    initial_future = add(1, 2)
+    futures = [initial_future]
+
+    if inline:
+        futures.append(add_inline(initial_future, 3))
+
+    if nested:
+        futures.append(add4_nested(initial_future, 1, 2, 3))
+
+    if fan_out > 0:
+        futures.append(add_fan_out(initial_future, fan_out))
+
+    if should_raise:
+        futures.append(do_raise(initial_future))
+
+    if raise_retry_probability:
+        futures.append(do_retry(initial_future))
+
+    if oom:
+        futures.append(do_oom(initial_future))
```

Comment:
It would be nice to have a configurable sleep as well, to test things that might take a while to hit. The one with the configurable sleep can also put out a log message every second, to help test logging with more than one page of logs.

Reply:
Done.